Civi/Core/Format.php

```diff
@@ -0,0 +1,178 @@
+  public function moneyNumber($amount, string $currency, $locale): string {
+    $formatter = $this->getMoneyFormatter($currency, $locale, NumberFormatter::DECIMAL);
+    $money = Money::of($amount, $currency, NULL, RoundingMode::HALF_UP);
+    return $money->formatWith($formatter);
+  }
+
+  /**
+   * Get a money value with formatting but not rounding.
+   *
+   * @param string|float|int $amount
+   * @param string|null $currency
+   * @param string|null $locale
+   *
+   * @return string
+   *
+   * @noinspection PhpDocMissingThrowsInspection
+   * @noinspection PhpUnhandledExceptionInspection
+   */
+  public function moneyLong($amount, ?string $currency, ?string $locale): string {
+    $formatter = $this->getMoneyFormatter($currency, $locale, NumberFormatter::CURRENCY, [
+      NumberFormatter::MAX_FRACTION_DIGITS => 9,
+    ]);
+    $money = Money::of($amount, $currency, new AutoContext());
+    return $money->formatWith($formatter);
+  }
+
+  /**
+   * Get a number with minimum decimal places based on the currency but no rounding.
+   *
+   * @param string|float|int $amount
+   * @param string|null $currency
+   * @param string|null $locale
+   *
+   * @return string
+   *
+   * @noinspection PhpDocMissingThrowsInspection
+   * @noinspection PhpUnhandledExceptionInspection
+   */
+  public function moneyNumberLong($amount, ?string $currency, ?string $locale): string {
+    $formatter = $this->getMoneyFormatter($currency, $locale, NumberFormatter::DECIMAL, [
+      NumberFormatter::MAX_FRACTION_DIGITS => 9,
+    ]);
+    $money = Money::of($amount, $currency, new AutoContext());
+    return $money->formatWith($formatter);
+  }
+
+  /**
+   * Should we use the configured thousand & decimal separators.
+   *
+   * The goal is to phase this into being FALSE - but for now
+   * we are looking at how to manage an 'opt in'
+   */
+  protected function isUseSeparatorSettings(): bool {
+    return !CRM_Utils_Constant::value('IGNORE_SEPARATOR_CONFIG');
+  }
+
+  /**
+   * Get the money formatter for when we are using configured thousand separators.
+   *
+   * Our intent is to phase out these settings in favour of deriving them from the locale.
+   *
+   * @param string|null $currency
+   * @param string|null $locale
+   * @param int $style
+   *   See https://www.php.net/manual/en/class.numberformatter.php#intl.numberformatter-constants
+   * @param array $attributes
+   *   See https://www.php.net/manual/en/class.numberformatter.php#intl.numberformatter-constants.unumberformatattribute
+   *
+   * @return \NumberFormatter
+   *
+   * @noinspection PhpDocMissingThrowsInspection
+   * @noinspection PhpUnhandledExceptionInspection
+   */
+  public function getMoneyFormatter(?string $currency = NULL, ?string $locale = NULL, int $style = NumberFormatter::CURRENCY, array $attributes = []): NumberFormatter {
+    if (!$currency) {
+      $currency = Civi::settings()->get('defaultCurrency');
+    }
+    $formatter = new NumberFormatter($locale, $style);
```

Comment:
A local cache (`$this->formatters[$locale . $currency]` or similar) is probably worthwhile here... But you can check the numbers.

I ran a basic benchmark - [generate ~1000 example numbers and format them](https://gist.github.com/totten/7b45fef1ab81eafb032d006213b49f79) with slight variations in the formatting strategy. On a relatively fast laptop (i7-9750H; no other major tasks) , this pretty consistently returned runtimes like so:

```
## Baseline/Control: use sprintf() with fixed length
staticPrintf: 0.001189 for 1002 amounts

## Use a single, static NumberFormatter - disregarding the variations in locale and currency
staticFmt: 0.062295 for 1002 amounts

## Create new NumberFormatter each time you format number
newFmt: 0.124439 for 1002 amounts across 3 locale-currency context(s)

## Create a few NumberFormatters, cache+reuse them
cachedFmt: 0.062991 for 1002 amounts across 3 locale-currency context(s)
```

For ~1,000 numbers, the `newFmt` is +120ms and the `cachedFmt` is +60ms.

I don't think it's a stretch to see ~1,000 numbers in one request - e.g. a data-table or spreadsheet summarizing contributions (with columns for total-amount, tax-amount, fee-amount, etc) would only need a couple hundred records. Granted, the more realistic target for a paginated data-table is more like `(50 rows x 5 columns)` - so only a quarter of that (say, 30ms vs 15ms). OTOH, the benchmark environment was rather favorable (wrt load-avg and CPU-type).

Anyway, that's not a hard block, but reusing the `NumberFormatter`s is a fairly simple thing and cuts the runtime-cost in half.

Reply:
@totten before I try to read ^^  - can you confirm what 'fmt' means?

Reply:
"format"

Reply:
OK - I've pushed up a caching fix

Reply:
I'll try it in the benchmark.

Reply:
Updated benchmark to be more complete/realistic. It now runs literal variations of `Civi\Core\Format`. https://gist.github.com/totten/6c2759bfc661b99a6ec594d752f9e782

I ran it with `$format->number($amount, $locale)` and `$format->money($amount, $currency, $locale)`. The metadata-cache doesn't work (and doesn't make sense to me), so I threw converted it to `static-cache`. The names now correspond to the file being benchmarked.

| Operation | dummy-sprintf | singleton | new-inst | static-cache | static-cache-less-md5 |
| -- | -- | -- | -- | -- | -- |
| `number()` | 0.001 | 0.020 | 0.068 | 0.054 | 0.051 |
| `money()` | 0.001 | 0.319 | 0.390 | 0.331 | 0.315 |

Comparing `new-inst.php` and `static-cache-less-md5.php` in this benchmark... the gain for `number()` isn't quite as much as I'd expected from the more basic benchmark, but it is there (68ms vs 50ms). The gain for `money()` is a bit more notable (390ms vs 315ms).

```
=== benchmark of `number($amount, $locale)` ===

## Format all numbers with a dummy sprintf() call
Load: /Users/totten/bknix/build/dmaster/web/sites/all/modules/civicrm/tmp/bench/dummy-sprintf.php
dummy-sprintf.php: Warmup: 0.000003
dummy-sprintf.php: Runtime: 0.001531 for 1002 amounts across 3 locale-currency context(s)

## Format all numbers with the same single instance of NumberFormatter
Load: /Users/totten/bknix/build/dmaster/web/sites/all/modules/civicrm/tmp/bench/singleton.php
singleton.php: Warmup: 0.005992
singleton.php: Runtime: 0.020986 for 1002 amounts across 3 locale-currency context(s)

## Make a new instance of NumberFormatter for each number
Load: /Users/totten/bknix/build/dmaster/web/sites/all/modules/civicrm/tmp/bench/new-inst.php
new-inst.php: Warmup: 0.005537
new-inst.php: Runtime: 0.068378 for 1002 amounts across 3 locale-currency context(s)

## Cache NumberFormater in Civi::cache('metadata')
Load: /Users/totten/bknix/build/dmaster/web/sites/all/modules/civicrm/tmp/bench/metadata-cache.php
  [Symfony\Component\Debug\Exception\FatalThrowableError]
  Type error: Return value of Brick\Money\Money::formatWith() must be of the type string, boolean returned

## Cache NumberFormater in Civi::$statics
Load: /Users/totten/bknix/build/dmaster/web/sites/all/modules/civicrm/tmp/bench/static-cache.php
static-cache.php: Warmup: 0.008279
static-cache.php: Runtime: 0.054834 for 1002 amounts across 3 locale-currency context(s)

## Cache NumberFormater in Civi::$statics, but use trick to reduce `md5(json_encode(...))` work.
Load: /Users/totten/bknix/build/dmaster/web/sites/all/modules/civicrm/tmp/bench/static-cache-less-md5.php
static-cache-less-md5.php: Warmup: 0.007369
static-cache-less-md5.php: Runtime: 0.051832 for 1002 amounts across 3 locale-currency context(s)

=== benchmark of `money($amount, $currency, $locale)`

## Format all numbers with a dummy sprintf() call
Load: /Users/totten/bknix/build/dmaster/web/sites/all/modules/civicrm/tmp/bench/dummy-sprintf.php
dummy-sprintf.php: Warmup: 0.000003
dummy-sprintf.php: Runtime: 0.001588 for 1002 amounts across 3 locale-currency context(s)

## Format all numbers with the same single instance of NumberFormatter
Load: /Users/totten/bknix/build/dmaster/web/sites/all/modules/civicrm/tmp/bench/singleton.php
singleton.php: Warmup: 0.013554
singleton.php: Runtime: 0.319169 for 1002 amounts across 3 locale-currency context(s)

## Make a new instance of NumberFormatter for each number
Load: /Users/totten/bknix/build/dmaster/web/sites/all/modules/civicrm/tmp/bench/new-inst.php
new-inst.php: Warmup: 0.005445
new-inst.php: Runtime: 0.390538 for 1002 amounts across 3 locale-currency context(s)

## Cache NumberFormater in Civi::cache('metadata')
Load: /Users/totten/bknix/build/dmaster/web/sites/all/modules/civicrm/tmp/bench/metadata-cache.php
  [Symfony\Component\Debug\Exception\FatalThrowableError]
  Type error: Return value of Brick\Money\Money::formatWith() must be of the type string, boolean returned

## Cache NumberFormater in Civi::$statics
Load: /Users/totten/bknix/build/dmaster/web/sites/all/modules/civicrm/tmp/bench/static-cache.php
static-cache.php: Warmup: 0.007789
static-cache.php: Runtime: 0.331434 for 1002 amounts across 3 locale-currency context(s)

## Cache NumberFormater in Civi::$statics, but use trick to reduce `md5(json_encode(...))` work.
Load: /Users/totten/bknix/build/dmaster/web/sites/all/modules/civicrm/tmp/bench/static-cache-less-md5.php
static-cache-less-md5.php: Warmup: 0.009057
static-cache-less-md5.php: Runtime: 0.315846 for 1002 amounts across 3 locale-currency context(s)
```

Reply:
@totten ah - we can't cache an object. Ah well using array-backed Redis caching makes a difference on single page loads but not batch jobs anyway

I guess it would be interesting to compare against the existing Money::format function - but we probably have enough info

Reply:
Yeah, I think a static-cache is low-hanging-fruit.

I'm not certain what an apples-to-apples benchmark would be for `CRM_Utils_Money::format()`. I took [a quick stab](https://gist.github.com/totten/6c2759bfc661b99a6ec594d752f9e782#file-crm-utils-php) and ran a few times locally - for `run_benchmark_money()`, it seems to consistently come in between (*`static-cache.php` beats `crm-utils.php` which beats `new-inst.php`*). For `run_benchmark_number()`, it comes in as the slowest of the pack --but it may just be that I didn't find the right operation to compare against.

Reply:
Well I flipped it to Civi::statics